### PR TITLE
UCS/TOPO: distance query based on dev-path difference - v1.9.x

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What
Backport of https://github.com/openucx/ucx/pull/5339

### CODE REVIEW

**Comment by Akshay-Venkatesh** [comment]
Doesn't look like a related error
```
[----------] 1 test from log_test_file_size, where TypeParam = 
[ RUN      ] log_test_file_size.large_files <> <>
[     INFO ] "grep 'Z1s5frVqFZpPRPtV13wOhYUmzHOBY0461u9fmdVzCIMbgG7fH13' /tmp/gtest_ucs_log.10471.*" failed after 20 iterations (return value: 127)
/scrap/azure/agent-03/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucs/test_log.cc:221: Failure
Value of: do_grep(test_str)
  Actual: false
Expected: true
[     INFO ] "grep '664YQlEu2RIMrWiD4WRBnNvKK2WLcsYjybFMMTGMuXik3Ox5vOGgkBy2BVNoNtVJUido9tA3OqnSDjV85mmpvkp7DAta4MrW4UJDMTEkI0AVjgboQNcL5PQJaHrcdh8' /tmp/gtest_ucs_log.10471.*" failed after 20 iterations (return value: 127)
/scrap/azure/agent-03/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucs/test_log.cc:221: Failure
Value of: do_grep(test_str)
  Actual: false
Expected: true
[  FAILED  ] log_test_file_size.large_files, where TypeParam =  and GetParam() =  (8289 ms)
[----------] 1 test from log_test_file_size (8289 ms total)
``` 
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=7702&view=logs&j=88d29b50-4e11-53d5-399b-e634502852e6&t=cb773578-1ba5-52b8-cb10-437b22e73aab

**Comment by brminich** [comment]
issue is #5241

**Comment by brminich** [comment]
ok to test

File: src/ucs/sys/string.c [hunk 1]
```diff
@@ -293,3 +294,34 @@ const char* ucs_flags_str(char *buf, size_t max,
 
     return buf;
 }
+
+ssize_t ucs_path_calc_distance(const char *path1, const char *path2)
+{
+    unsigned distance = 0;
+    int same  = 1;
```
Comment (dmitrygx) [inline src/ucs/sys/string.c:301]: [minor] unaligned with the line above

File: src/ucs/sys/string.c [hunk 2]
```diff
@@ -293,3 +294,34 @@ const char* ucs_flags_str(char *buf, size_t max,
...
+    int comp_len;
+    int i;
+    char resolved_path1[PATH_MAX], resolved_path2[PATH_MAX];
+    int rp_len1, rp_len2;
```
Comment (dmitrygx) [inline src/ucs/sys/string.c:305]: `size_t`? since strlen return `size_t`
and
`size_t i, comp_len;`

File: test/gtest/ucs/test_topo.cc [hunk 3]
```diff
@@ -39,7 +39,7 @@ UCS_TEST_F(test_topo, get_distance) {
 
     status = ucs_topo_get_distance(UCS_SYS_DEVICE_ID_UNKNOWN,
                                    UCS_SYS_DEVICE_ID_UNKNOWN, &distance);
-    ASSERT_UCS_OK(status);
+    ASSERT_EQ(status, UCS_ERR_IO_ERROR);
```
Comment (dmitrygx) [inline test/gtest/ucs/test_topo.cc:42]: ASSERT_EQ(UCS_ERR_IO_ERROR, status);
to correctly report `expected, actual` in the result

**Pull request merged**